### PR TITLE
feature/betalist

### PR DESCRIPTION
**Pull request by hakimelek**
@lorensr 

* Creates a email subscribe mutation using Apollo Server and MongoDB.
* Adds a new React Form Component. 
* Handles form validation. 

TODO: 
- [ ] React Component Tests.
- [ ] Mutation Tests.
- [ ] Better component styling.

### CODE REVIEW

File: data/resolvers.js [hunk 1]
```diff
@@ -1,7 +1,61 @@
...
+            reason: 'A subscriber with this email already exists',
+          },
+        });
+      } else if (args.email.length === 0) {
+        throw new DbError({
+          data: {
+            reason: 'The email provided should not be empty',
+          },
+        });
+      } else if (!/^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/.test(args.email)) {
```
Comment (lorensr) [inline data/resolvers.js]: is this kind of test being done client-side as well?

File: data/resolvers.js [hunk 2]
```diff
@@ -1,7 +1,61 @@
...
+        throw new DbError({
+          data: {
+            reason: 'Provide a valid email',
+          },
+        });
+      }
```
Comment (lorensr) [inline data/resolvers.js]: shorter would be 
```
let error
if ()
  error = "A subscriber"
else if 
  error = ...

if (error) {
 throw new
    reason: error
```

File: data/resolvers.js [hunk 3]
```diff
@@ -1,7 +1,61 @@
...
+      const subscriber = new Subscriber({
+        email: args.email,
+      });
+
+      try {
+        return await subscriber.save();
+      } catch (err) {
+        throw err;
+      }
+    }
```
Comment (lorensr) [inline data/resolvers.js]: isn't this equivalent to just `return await subscriber.save()`?

File: data/schema.js [hunk 4]
```diff
@@ -1,9 +1,19 @@
 export default [`
+type Subscriber {
+  email: String
+}
+
 type Query {
-  foo: String
+  subscriber (email: String!): Subscriber
+  allSubscribers: [Subscriber]
+}
```
Comment (lorensr) [inline data/schema.js]: don't think queries are needed? and re: second one, don't want anyone to be able to get my list of subscribers

File: lib/initClient.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+let apolloClient = null
+
+function createClient (headers) {
+  return new ApolloClient({
+    ssrMode: !process.browser,
+    headers,
+    dataIdFromObject: result => result.id || null,
+    networkInterface: createNetworkInterface({
+      uri: 'http://localhost:3000/graphql',
```
Comment (lorensr) [inline lib/initClient.js]: can this arg be relative to current domain? like would '/graphql' work?

**Comment by lorensr** [comment]
bug: when hitting enter on form, page reloads 😁

**Comment by hakimelek** [comment]
Will fix that!

File: lib/initClient.js (hunk 5 above)
Comment (hakimelek) [inline lib/initClient.js]: window.location.host, window would throw an error, maybe because of SSR. What do you suggest?

**Comment by lorensr** [comment]
check out last commit. 

also, want to try this?

http://stackoverflow.com/questions/33211672/how-to-submit-a-form-using-enter-key-in-react-js

File: data/resolvers.js (hunk 3 above)
Comment (lorensr) [inline data/resolvers.js]: but when you catch, you throw. just throwing whatever you catch is equivalent to not catching in the first place

File: data/resolvers.js (hunk 1 above)
Comment (lorensr) [inline data/resolvers.js]: no, it's to get a faster feedback to the user. it's fast for us in development, but in production will require a round-trip to the server

**Comment by hakimelek** [comment]
Thanks for the formatting :)

**Comment by lorensr** [comment]
Deployed to: https://graphql.guide/

Permanent link: https://graphql-guide-ebkblcfhbv.now.sh/

**Pull request merged**